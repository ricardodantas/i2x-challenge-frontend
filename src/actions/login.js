import { createAction } from 'redux-actions'

export const loginFetching = createAction('LOGIN_FETCHING')
export const loginError    = createAction('LOGIN_ERROR')
export const loginSuccess  = createAction('LOGIN_SUCCESS')
export const logout = createAction('LOGOUT')

export function doLogout() {
  return sessionStorage.removeItem('token')
}

export function login(email, password) {
  return dispatch => {
    dispatch(loginFetching())
    return fetch('https://i2x-challenge.herokuapp.com/core/login/',{
        method: 'POST',
        mode: 'cors',
        cache: 'no-cache',
        headers: new Headers(
           {
             "content-type": "application/json",
          }
        ),
        body: JSON.stringify(
          {
            email: email,
            password: password
          }
        )
     })
    .then((response) => response.ok ? response.json() : Promise.reject(response.text()))
    .then((responseJson) => dispatch(loginSuccess(responseJson)))
    .catch((error) => dispatch(loginError(error.message)))
  }
}
