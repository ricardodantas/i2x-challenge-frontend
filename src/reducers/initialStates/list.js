export const initialState = {
  isFetching: false,
  hasFailed: false,
  hasSuccess: false,
  items: [],
}
