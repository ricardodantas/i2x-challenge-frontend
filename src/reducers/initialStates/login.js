import {isEmpty} from 'lodash'
import apiAuth from '../../utils/auth'

// @flow
export const initialState: {
  isFetching: boolean,
  hasFailed: boolean,
  hasSuccess: boolean,
  response: Object<null>,
} = {
  isAuthenticated: apiAuth.isAuthenticated(),
  isFetching: false,
  hasFailed: false,
  hasSuccess: false,
  response: {},
}
