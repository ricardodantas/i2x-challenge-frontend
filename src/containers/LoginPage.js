import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {Redirect} from 'react-router-dom'
import { connect } from 'react-redux'
import LoginForm from '../components/LoginForm'
import {login} from '../actions/login'
import { Dimmer, Loader, Header } from 'semantic-ui-react'

class LoginPage extends Component {

  static propTypes = {
    login: PropTypes.func.isRequired,
    requestLogin: PropTypes.object.isRequired,
  }

  static defaultProps = {
    requestLogin: {}
  }

  state = {
    email: '',
    password: '',
  }

  handleChange = (e, { name, value }) => this.setState({ [name]: value })

  handleSubmit = e => {
    e.preventDefault()

    const {dispatch, login } = this.props
    const { email, password } = this.state

    dispatch(login(email, password))
  }

  render() {
    const {requestLogin, location} = this.props
    const {isAuthenticated} = requestLogin

    if(isAuthenticated) {
      return <Redirect to="/"/>
    }

    return (
      <div>
        <Header as='h2' icon textAlign='center'>
          <Header.Content>
            i2x Challenge Frontend
          </Header.Content>
        </Header>
        <LoginForm response={requestLogin} handleChange={this.handleChange} handleSubmit={this.handleSubmit}/>
        <Dimmer active={requestLogin.isFetching}>
          <Loader size="massive" active={requestLogin.isFetching}/>
        </Dimmer>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    login,
    dispatch
  }
}

const mapStateToProps = (state) => {
  return {
    requestLogin: state.login
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage)
