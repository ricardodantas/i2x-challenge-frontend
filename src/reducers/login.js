import { handleActions } from 'redux-actions'
import {login, logout, doLogout, loginFetching, loginSuccess, loginError} from '../actions/login'
import _ from 'lodash'

const initialState = {
  isAuthenticated: !_.isEmpty(sessionStorage.getItem('token')),
  isFetching: false,
  hasFailed: false,
  hasSuccess: false,
  token: null,
  response: {},
}

export default handleActions({

  [logout]: (state, action) => {
    return Object.assign({}, state, {
      isAuthenticated: false,
      token: null,
      response: {},
    })
  },
  [loginFetching]: (state, action) => {
    return Object.assign({}, state, {
      isAuthenticated: false,
      isFetching: true,
      hasSuccess: false,
      hasFailed: false,
      token: null,
      response: action
    })
  },
  [loginSuccess]: (state, action) => {
    return Object.assign({}, state, {
      isAuthenticated: true,
      isFetching: false,
      hasSuccess: true,
      token: action.payload.token,
      response: action
    })
  },
  [loginError]: (state, action) => {
    return Object.assign({}, state, {
      isFetching: false,
      hasFailed: true,
      response: action
    })
  },
}, initialState)
