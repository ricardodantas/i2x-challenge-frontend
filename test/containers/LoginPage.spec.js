import React from 'react'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import LoginPage from '../../src/containers/LoginPage'
import configureStore from '../../src/store/configureStore'
import {initialState} from '../../src/reducers/initialStates/login'

describe('<LoginPage />', () => {


  beforeEach(() => {
    global.store = configureStore({ login: initialState })
  })

  it('should display login page', () => {

    const tree = renderer.create(
      <Provider store={global.store}>
        <LoginPage />
      </Provider>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should display LoginForm component', () => {
    const component = mount(
      <Provider store={global.store}>
        <LoginPage />
      </Provider>
    )

    const loginForm = component.find('LoginForm')
    expect(loginForm.exists()).toBe(true)
  })

  it('should display Header component', () => {
    const component = mount(
      <Provider store={global.store}>
        <LoginPage />
      </Provider>
    )

    const header = component.find('Header')
    expect(header.exists()).toBe(true)
  })

})
