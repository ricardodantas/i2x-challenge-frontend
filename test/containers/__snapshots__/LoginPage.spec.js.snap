// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> should display login page 1`] = `
<div>
  <h2
    className="ui icon center aligned header"
  >
    <div
      className="content"
    >
      Sample React, Redux, Rouer and JWT Auth
    </div>
  </h2>
  <div
    className="ui container ui raised very padded text container segment"
  >
    <h2>
      Login
    </h2>
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <div
          className="ui input"
        >
          <input
            name="email"
            onChange={[Function]}
            placeholder="email"
            type="email"
          />
        </div>
      </div>
      <div
        className="field"
      >
        <div
          className="ui input"
        >
          <input
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
          />
        </div>
      </div>
      <button
        className="ui primary button"
        onClick={[Function]}
        tabIndex={undefined}
        type="submit"
      >
        Sign In
      </button>
    </form>
    <div
      className="ui hidden negative message"
    >
      <div
        className="header"
      >
        Ops...
      </div>
      <p>
        Something is wrong.
      </p>
    </div>
  </div>
  <div
    className="ui dimmer"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div
        className="center"
      >
        <div
          className="ui massive loader"
        />
      </div>
    </div>
  </div>
</div>
`;
